indexing-service/src/main/java/io/druid/indexing/common/task/SinglePhaseParallelIndexTaskRunner.java

```diff
@@ -0,0 +1,624 @@
+              throw new ISE("spec[%s] is in an invalid state[%s]", taskCompleteEvent.getSpec().getId(), completeState);
+          }
+        }
+      }
+    }
+    finally {
+      log.info("Cleaning up resources");
+      // Cleanup resources
+      taskCompleteEvents.clear();
+      taskMonitor.stop();
+      chatHandlerProvider.unregister(taskId);
+
+      if (state != TaskState.SUCCESS) {
+        log.info(
+            "This task is finished with [%s] state. Killing [%d] remaining subtasks.",
+            state,
+            taskMonitor.getNumRunningTasks()
+        );
+        // if this fails, kill all sub tasks
+        // Note: this doesn't work when this task is killed by users. We need a way for gracefully shutting down tasks
+        // for resource cleanup.
+        taskMonitor.killAll();
+      }
+    }
+
+    return state;
+  }
+
+  @VisibleForTesting
+  ParallelIndexIngestionSpec getIngestionSchema()
+  {
+    return ingestionSchema;
+  }
+
+  private String getDataSource()
+  {
+    return ingestionSchema.getDataSchema().getDataSource();
+  }
+
+  // Internal API for collecting reports from subTasks
+
+  @POST
+  @Path("/report")
+  @Consumes(SmileMediaTypes.APPLICATION_JACKSON_SMILE)
+  public Response report(
+      PushedSegmentsReport report,
+      @Context final HttpServletRequest req
+  )
+  {
+    ChatHandlers.authorizationCheck(
+        req,
+        Action.WRITE,
+        getDataSource(),
+        authorizerMapper
+    );
+    collectReport(report);
+    return Response.status(Response.Status.OK).build();
+  }
+
+  @VisibleForTesting
+  void collectReport(PushedSegmentsReport report)
+  {
+    // subTasks might send their reports multiple times because of the HTTP retry.
+    // Here, we simply make sure the current report is exactly same with the previous one.
+    segmentsMap.compute(report.getTaskId(), (taskId, prevReport) -> {
+      if (prevReport != null) {
+        Preconditions.checkState(
+            prevReport.getSegments().equals(report.getSegments()),
+            "task[%s] sent two or more reports and previous report[%s] is different from the current one[%s]",
+            taskId,
+            prevReport,
+            report
+        );
+      }
+      return report;
+    });
+  }
+
+  private void publish(TaskToolbox toolbox) throws IOException
+  {
+    final TransactionalSegmentPublisher publisher = (segments, commitMetadata) -> {
+      final SegmentTransactionalInsertAction action = new SegmentTransactionalInsertAction(segments);
+      return toolbox.getTaskActionClient().submit(action).isSuccess();
+    };
+    final UsedSegmentChecker usedSegmentChecker = new ActionBasedUsedSegmentChecker(toolbox.getTaskActionClient());
+    final Set<DataSegment> segmentsToPublish = segmentsMap
+        .values()
+        .stream()
+        .flatMap(report -> report.getSegments().stream())
+        .collect(Collectors.toSet());
+    final boolean published = publisher.publishSegments(segmentsToPublish, null);
```

Comment:
(probably out of scope for this PR) being able to pass in the `Stream<DataSegment>` directly here would be nice

Reply:
Hmm, `segmentsToPublish` is used to check all segments are successfully published like below:

```java
      if (usedSegmentChecker.findUsedSegments(segmentsIdentifiers)
                            .equals(segmentsToPublish)) {
        log.info("Our segments really do exist, awaiting handoff.");
      } else {
        throw new ISE("Failed to publish segments[%s]", segmentsToPublish);
      }
```

I think it should be materialized anyway, so we don't have to do that if we can get a lot of benefits.